### ISSUE TITLE
network type is unclear

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#419]
Originally reported of Google Code with ID 363.

```
The network type can be public, private, or None.
Neither the design nor more importantly the manpage though specify what it means for a network to be public or private or not have a type.
This should be documented.
```
Originally added on 2013-02-07 13:07:58 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
As discussed on the list, I strongly object against "None" (the Python value, not the string) as a value for this type. If there must be a type outside private/public, then make it a standalone real value, e.g. "default".
```
Originally added on 2013-02-07 13:11:51 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
+1 against the python value. For me "none" the string or "default" doesn't change much, until we know what it is and what we should do with it.
If ganeti doesn't use it at all maybe we should drop it completely in favor of tags? I'm waiting for discussion. :)

Thanks!

Guido

```
Originally added on 2013-02-07 13:26:02 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I agree with dropping that slot completely in favor of tags.
Ganeti does not use it at all currently (nor will in the future). Only ifup scripts
use it. So having that kind of info in tags suits best. I'll fix it
as soon as possible.

Thanks,
dimara
```
Originally added on 2013-02-07 14:58:15 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Thanks, please send a patch to devel-2.7, making sure if possible to delete the existing slot, if found, when upgrading the object (see how other objects in objects.py do it). :) Should "family" have the same fate? (see the other bug I submitted). 

Thanks!

Guido

```
Originally added on 2013-02-08 07:26:55 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
In particular since the slot is there in beta1, we should leave the slot there and just force-set it to None for the whole beta series (and of course not export it in hooks and such). We can then drop it once 2.7 is released, so that upgrading during the beta series doesn't break compatibility. (we won't support upgrading directly from beta1 to 2.8)
```
Originally added on 2013-02-08 07:49:15 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Other question (sorry for them coming one at a time): is "size" in the same situation, and should be removed? That one seems not to ever even be set anywhere in the code.

Thanks,

Guido
```
Originally added on 2013-02-08 07:55:36 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
> In particular since the slot is there in beta1, we should leave the slot there and just force-set it to None for the whole beta series (and of course not export it in hooks and such).

I'm not 100% what you propose, but if you propose to have this field in the stable 2.7 series, I disagree. If this would have been in RC, maybe, but in beta? We don't support betas as "upgradable". Just change the code to drop it, and let's have 2.7 clean as well.

Otherwise people might start to use it in 2.7 already. Or, alternatively, we would have to announce "we have this field, new in 2.7, but don't use it, because it's already deprecated".
```
Originally added on 2013-02-08 09:05:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
iustin: I'm just proposing not to drop the "slot", not the "field". I am happy to drop it after beta2, or rc1, no problem at all, I'd just keep it at least "one tarball more": note that having "just" the slot if UpgradeConfig sets it to None doesn't make it usable at all.
```
Originally added on 2013-02-08 10:16:52 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Just to clarify something discussed internally. Guido proposed to leave the field in order to deprecate it via "Upgrade***()" method. My proposal was to add a custom "FromDict" method for Network so that we can do this:

@@ -2037,6 +2037,16 @@ class Network(TaggableObject):
     "ext_reservations",
     ] + _TIMESTAMPS + _UUID
 
+  @classmethod
+  def FromDict(cls, val):
+    """Custom function for networks.
+
+    """
+    if "network_type" in val:
+      del val["network_type"]
+    obj = super(TaggableObject, cls).FromDict(val)
+    return obj
+

(Untested, beware, etc.)
```
Originally added on 2013-02-08 10:28:09 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Well I am not sure what way to go. I got confused. The certain part is that all the
code referring to network_type in cmdlib/config/gnt_network/opcodes etc. should go away. OK.

The slot in objects.py should be removed?
If yes it breaks existing configuration data. Workarounds are the following:
 - Add a function in cfgupgrade (e.g DeleteDeprecatedNetworkType())
 - Add a FromDict in Network object
 - Add a UpgradeConfig in Network Object.

How about runtime files? They contain netinfo in every NIC (which sould be removed
in a different patch). I think FromDict workaround is the only one that doesn't break in this case.

If the slot will not be removed then there is no reason to add any of these workarounds. Just remove the code that exposes it to the end user.

Do I stand correctly? Which way to go?

dimara 
```
Originally added on 2013-02-08 16:15:05 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Please go with dropping the slot and adding the FromDict in the Network object. :)
(the rest of the considerations I don't fully agree with, but it doesn't matter as we have a preferred way to go) :)

Thanks,

Guido
```
Originally added on 2013-02-08 16:29:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-02-11 15:32:07 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
